Trust domain intro 00002 (#3)
enarx/rfcs: rfc#00002.md

# RFC00002: Trust domain introduction
- Authors: [Mike Bursell](mike@p2ptrust.org)
- Status: [PROPOSED](/README.md#proposed)
- Since: 2020-02-13
- Status Note: under discussion
- Start Date: 2020-02-11
- Tags: trust

## Summary

Trust domains are a key concept within Enarx.  This RFC provides an
introduction to trust domains.

## Motivation

In order to explain the motivations behind Enarx, our design principles
and aims, we need to be able to explain what trust domains are.  This RFC
attempts to provide an introduction.

## Tutorial

Before considering trust, we need to consider why it is relevant
to the issue of TEEs and Enarx.  The fundamental concept behind

> Enarx is that it allows applications - workloads - to be run on

Reviewer: ~~applications - workloads - to be~~ -> applications, also referred to as "workloads", to be

Author: fixed.

> machines - hosts - which are not under the control of the owner

Reviewer: ~~machines - hosts - which are not under the control of the owner of the workload~~ -> hosts which are not under the tenant's control

Author: No - there are sometimes distinctions to be made here, particularly if we're looking at non-Cloud scenarios such as Edge or IoT.  I've added a little text.

> of the workload.  The workload owner does not trust the host or

Reviewer: ~~workload owner~~ -> tenant

Author: addressed elsewhere, following suggestions from @axelsimon

the host owner.  But does this mean?

Trust, in the human sphere, can be thought of the expectation one

> person - the trustor - has that another person - the trustee - will

Reviewer: ~~person - the trustor - has that another person - the trustee - will~~ -> person, the trustor, has that another person, the trustee, will

Author: Nope.

Author: No.

behave in a particular way.  The mapping to the computing domain
is not perfect, but if we want to think about the relevant human
actors, we can model of the trustor as the workload owner
and the host owner as trustee.  In order to map this more directly
to the realm of computer systems, we can say that the process
deploying the workload is the trustor, and that the host is the
trustee.

How can the deploying process trust the host to execute the
workload as expected, and what do we mean by "expected"?  The
specific expectations in which we are interested are the
confidentiality of the workload and the integrity of the workload.
In the normal computing case, there can be no checkable or
cryptographically attested assurance tha tthe host will meet
these expectations.  Trusted Execution Environments provide

> ways to build such assurances, and for trust to be built in such

Reviewer 2: ```suggestion
these expectations.  Trusted Execution Environments provide
means to build such assurances, and for trust to be built in such
```

Author: Thanks.

a way that the deploying process (which represents, or has agency
for, the workload owner) can have assurances that the host will
not - and cannot - break confidentiality or integrity of a
workload which it (the host) is executing.

A definition of trust, from [What is trust?](http://aliceevebob.com/2017/05/09/what-is-trust-with-apologies-to-pontius-pilate/).

"Trust is the assurance that one entity holds that another will
perform particular actions according to a specific expectation."
First corollary: "Trust is always contextual."
Second corollary: "One of the contexts for trust is always time."
Third corollary: "Trust relationships are not symmetrical."

#### Applying this to Enarx
The key trust relationship for Enarx is that held by the tenant

> (workload owner) to the workload it owns.  Enarx aims to provide assurances

Reviewer 2: "tenant (workload owner)": would it be worth rationalising the various ways tenant and workload owner are put together in the RFC? Most of the time, workload owner comes first, with tenant inside dashes or comas, here it is reversed, and uses parentheses. A minor squabble, definitely, but maybe worth addressing to make reading easier.
Another way to simplify would be to clearly equate "workload owner" with tenant early on in the text, and then only refer to tenant afterwards.

Author: Hopefully fixed.

to the tenant that the integrity and confidentiality of that workload are
protected: these are the key contexts of that trust, and actions can be
assured if they are maintained.  Trust may decay or be built up in the
workload - this decay or building represents the time context.  In this
case, the workload does not have an equivalent set of assurances about
the tenant's integrity and confidentiality, so the relationship if
there were one, would not be symmetrical.  For our purposes, it is
the ability to assert and have assurances in regards to properties such
as these - integrity and confidentiality, both security properties -
that defines how we consider trust relationships.

### Trust domains
A trust domain can be defined as a set of entities or components that
share the same security policies.  These may be human or machine, but
machine entities are usually assumed, and are what are relevant to Enarx.
We can refer to these entities as members of a trust domain.  Trust
domains can be subsets or supersets of other trust domains, but may not
intersect: a true intersection between two trust domains implies that

> they share the same policies, which means that they are in union.

Reviewer 2: ```suggestion
they share the same policies, which means that they are in union, and thus
the same.
```
What do you think?

Author: Taken, parenthesised.

## Tutorial

Before considering trust, we need to consider why it is relevant
to the issue of TEEs and Enarx.  The fundamental concept behind
Enarx is that it allows applications - workloads - to be run on
machines - hosts - which are not under the control of the owner
of the workload.  The workload owner does not trust the host or

> the host owner.  But does this mean?

Reviewer 3: ```suggestion
the host owner.  But what does this mean?
```

Author: fixed

Trust, in the human sphere, can be thought of the expectation one
person - the trustor - has that another person - the trustee - will
behave in a particular way.  The mapping to the computing domain
is not perfect, but if we want to think about the relevant human
actors, we can model of the trustor as the workload owner
and the host owner as trustee.  In order to map this more directly
to the realm of computer systems, we can say that the process
deploying the workload is the trustor, and that the host is the
trustee.

How can the deploying process trust the host to execute the
workload as expected, and what do we mean by "expected"?  The
specific expectations in which we are interested are the
confidentiality of the workload and the integrity of the workload.
In the normal computing case, there can be no checkable or

> cryptographically attested assurance tha tthe host will meet

Reviewer 3: ```suggestion
cryptographically attested assurance that the host will meet
```

Author: fxd
